pkg/controllers/vdb/stopsubcluster_reconciler.go

```diff
@@ -0,0 +1,156 @@
+/*
+ (c) Copyright [2021-2024] Open Text.
+ Licensed under the Apache License, Version 2.0 (the "License");
+ You may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+*/
+
+package vdb
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	vapi "github.com/vertica/vertica-kubernetes/api/v1"
+	"github.com/vertica/vertica-kubernetes/pkg/cmds"
+	"github.com/vertica/vertica-kubernetes/pkg/controllers"
+	"github.com/vertica/vertica-kubernetes/pkg/events"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin"
+	"github.com/vertica/vertica-kubernetes/pkg/vadmin/opts/stopsc"
+	config "github.com/vertica/vertica-kubernetes/pkg/vdbconfig"
+	corev1 "k8s.io/api/core/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+)
+
+type StopSubclusterReconciler struct {
+	Rec            config.ReconcilerInterface
+	Vdb            *vapi.VerticaDB // Vdb is the CRD we are acting on.
+	PRunner        cmds.PodRunner
+	Log            logr.Logger
+	PFacts         *PodFacts
+	InitiatorPodIP string
+	Dispatcher     vadmin.Dispatcher
+}
+
+func MakeStopSubclusterReconciler(rec config.ReconcilerInterface, log logr.Logger,
+	vdb *vapi.VerticaDB, pfacts *PodFacts, dispatcher vadmin.Dispatcher) controllers.ReconcileActor {
+	return &StopSubclusterReconciler{
+		Rec:        rec,
+		Log:        log.WithName("StopSubclusterReconciler"),
+		Vdb:        vdb,
+		PFacts:     pfacts,
+		Dispatcher: dispatcher,
+	}
+}
+
+func (s *StopSubclusterReconciler) Reconcile(ctx context.Context, _ *ctrl.Request) (ctrl.Result, error) {
+	err := s.PFacts.Collect(ctx, s.Vdb)
+	if err != nil {
+		return ctrl.Result{}, err
+	}
+
+	scsToStop := s.findSubclustersWithShutdownNeeded()
+	if len(scsToStop) == 0 {
+		s.Log.Info("Aborting stop subcluster. No subclusters needing shutdown were found.")
+		return ctrl.Result{}, nil
+	}
+	oldSize := len(scsToStop)
+	scsToStop = s.filterSubclustersWithRunningPods(scsToStop)
+	needRequeue := false
+	if len(scsToStop) != oldSize {
+		needRequeue = true
+	}
+
+	if err := s.stopSubclusters(ctx, scsToStop); err != nil {
```

Comment:
Stop subclusters should be done after requeue check(line 77). If one sc is not up before stopping subclusters, then you stop the other scs. We will come here again, and the needRequeue is still true. We will do another round of reconciliation.